lumos/optimal_control/scaled_mesh_ocp.py

```diff
@@ -633,19 +625,74 @@ def _set_scales(self, scales: Dict[str, Dict[str, float]]):
+        op = self.dec_var_operator
+
+        # First we create a dictioinary to summarise all scales in one place.
         var_scales = {
-            g: np.tile(1 / self.model.get_group_scales(g), (self.num_stages, 1))
-            for g in self.stage_var_groups
+            g: self.model.make_const_vector(g, 1.0) for g in self.model._implicit_inputs
         }
-        # TODO: We should set this somewhere else
         for var_name in self.global_var_names:
-            if var_name in scales:
-                var_scales[var_name] = scales[var_name]
-            else:
-                var_scales[var_name] = 1
-        self._dec_var_scales = self.dec_var_operator.flatten_var(**var_scales)
+            var_scales[var_name] = 1.0
+
+        # create default decision variable scales
+        self._dec_var_scales = np.ones(self.num_dec)
+
+        # Overwrite with scales from configs and update decision variable scales
+        for sc in scale_configs:
+            if isinstance(sc, GlobalVarScaleConfig):
+                var_scales[sc.name] = sc.value
+                self._dec_var_scales[op.get_global_var_index(sc.name)] = sc.value
+            elif isinstance(sc, StageVarScaleConfig):
+                var_scales[sc.group][
+                    op.get_var_index_in_group(sc.group, sc.name)
+                ] = sc.value
+                self._dec_var_scales[
+                    op.get_var_index_in_dec(sc.group, sc.name)
+                ] = sc.value
+
+        # Set constarint scales
+
+        continuity_scales = np.ravel(
+            np.tile(
+                var_scales["states"],
+                (self.num_intervals, op.num_stages_per_interval - 1, 1),
+            )
+        )
+        if self.is_condensed:
+            condensed_model_algebra_scales = np.concatenate(
+                [continuity_scales]
+                + [var_scales["con_outputs"], np.ones(self.model.num_residuals),]
+                * self.num_stages
+            )
+            self._constraints["model_algebra"].set_con_scales(
+                condensed_model_algebra_scales
+            )
+
+        else:
+            # NOTE: standard and condensed model_algebra constraints have the same scale.s
+            # HACK: hard-coded order here needs to correspond to order in implicit_con. How
+            # do we make it more robust?
+            # NOTE: here we scale the residuals with unity because:
+            # 1) they are not decision, we don't necessarily need to set their scales
+            # 2) the user are free to set the scales of the residual in the model.
+            model_algebra_scales = np.concatenate(
+                [
+                    var_scales["states"],
+                    var_scales["con_outputs"],
+                    np.ones(self.model.num_residuals),
+                ]
+                * self.num_stages
+            )
+            # HACK: we have an issue here that we want Convconstraints to automatically set
```

Comment:
this comment should probably go to ConvConstraints. And maybe we should give this another try?

Reply:
done